dashboard/app/assets/javascripts/client_state.js

```diff
@@ -42,6 +42,49 @@ dashboard.clientState.reset = function() {
 };
 
 /**
+ * Returns the client-cached copy of the level source for the given
+ * scriptLevelId, if it's newer than the given timestamp.
+ * @param {number} scriptLevelId
+ * @param {number=} timestamp
+ * @returns {string|undefined} Cached copy of the level source, or undefined if
+ *   the cached copy is missing/stale.
+ */
+dashboard.clientState.sourceForLevel = function (scriptLevelId, timestamp) {
+  var data = localStorage.getItem('source' + scriptLevelId);
+  if (data) {
+    try {
+      var parsed = JSON.parse(data);
+    } catch (e) {
+      return;
+    }
+    if (!timestamp || parsed.timestamp > timestamp) {
+      return parsed.source;
+    }
+  }
+};
+
+/**
+ * Cache a copy of the level source along with a timestamp. Posts to /milestone
+ * may be queued, so save the data in localStorage to present a consistent
+ * client view.
+ * @param scriptLevelId
+ * @param timestamp
+ * @param source
+ */
+dashboard.clientState.writeSourceForLevel = function (scriptLevelId, timestamp, source) {
```

Comment:
Would I ever make this call with a timestamp other than now? If not, remove timestamp as a param